### PR TITLE
OCPBUGS-38823: skip bundle images when computing the set of compatible architectures for a given pod

### PR DESCRIPTION
**Pull request by aleskandro**
This PR updates the registry inspector to handle bundle images correctly. Bundle images are intended to carry metadata-only information and not be multi-architecture.

It is possible for an operator bundle image to reference a valid operator supported by the cluster's architectures while the bundle image itself remains a single-architecture manifest image. Suppose the cluster (e.g., an arm64+ppc64le cluster) has no nodes supporting the bundle image's architecture (such as an amd64-built bundle). OLM may fail to run the job in that case because the node affinity settings could lead to an ImagePullBackOff error.

To address this, the GetCompatibleArchitectureSet function will now return the set of all the architectures supported by the operator (arm64, amd64, ppc64le, s390x) for OLM bundle images. If there are other container images specified in the pod (as in the OLM case), the node affinity will be set to the intersection of the architectures supported by the non-bundle images. As this fix is strictly related to the OLM bundles, we don't consider the case where a pod has only that bundle as a container image. In fact, the pod cannot run in such a case because no binaries are shipped in the image of a bundle to act as an entry point.

Testing: I added two integration tests for this. Having some e2e might be good (e.g., subscription-based operator installation, operator-sdk run, but I'd leave this choice to @lwan-wanglin and consider it a low-priority task for now as the scenarios we support consider mixarch amd64 + $second_arch today and this edge case would not be faced.

A manual test, though, could be done even before merging.

### CODE REVIEW

**Comment by aleskandro** [comment]
/override "Red Hat Konflux / multiarch-tuning-operator-enterprise-contract / multiarch-tuning-operator"

**Comment by aleskandro** [comment]
/override "Red Hat Konflux / multiarch-tuning-operator-enterprise-contract / multiarch-tuning-operator"

File: controllers/podplacement/pod_model.go [hunk 1]
```diff
@@ -227,6 +227,12 @@ func (pod *Pod) intersectImagesArchitecture(pullSecretDataList [][]byte) (suppor
 			log.V(3).Error(err, "Error inspecting the image", "imageName", imageName)
 			return nil, err
 		}
+		if err == nil && len(currentImageSupportedArchitectures) == 0 {
+			// This case happens when the image is an operator bundle image. We don't set any constraints based on the architecture they report.
+			// See https://issues.redhat.com/browse/OCPBUGS-38823 for more information.
+			log.V(4).Info("No supported architectures found for the image (This image should be a bundle).", "imageName", imageName)
```
Comment (prb112) [inline controllers/podplacement/pod_model.go]: per our discussion, this looks like a reasonable solution. wouldn't it be conversely reasonable to set all architectures? since it should work everywhere?

just a thought.

Comment (aleskandro) [reply controllers/podplacement/pod_model.go:233]: Hi Paul, I did this because "all architectures" is not really defined outside OCP.

But by doing that, we may remove this `if` statement that i didn't like either. 
I'm going to change it. Thanks for pointing this out :)

Comment (aleskandro) [reply controllers/podplacement/pod_model.go]: Hi Paul, I did this because "all architectures" is not really defined outside OCP.

But by doing that, we may remove this `if` statement that i didn't like either. 
I'm going to change it. Thanks for pointing this out :)

**Comment by aleskandro** [comment]
/override "Red Hat Konflux / multiarch-tuning-operator-enterprise-contract / multiarch-tuning-operator"

File: pkg/image/inspector.go [hunk 2]
```diff
@@ -115,33 +123,53 @@ func (i *registryInspector) GetCompatibleArchitecturesSet(ctx context.Context, i
...
+
 	config, err := parsedImage.OCIConfig(ctx)
+
 	if err != nil {
-		// Ignore errors due to invalid images at this stage
 		log.Error(err, "Error parsing the OCI config of the image")
 		return nil, err
 	}
-	return sets.New[string](config.Architecture), nil
+	if _, ok := config.Config.Labels["operators.operatorframework.io.metrics.builder"]; ok {
```
Comment (Prashanth684) [inline pkg/image/inspector.go]: have this as a const with a name similar to https://github.com/operator-framework/operator-sdk/blob/master/internal/annotations/metrics/metrics.go#L32 ?

Comment (aleskandro) [reply pkg/image/inspector.go]: Done

**Comment by aleskandro** [comment]
/override "Red Hat Konflux / multiarch-tuning-operator-enterprise-contract / multiarch-tuning-operator"

**Review by prb112 — COMMENTED** [review]
/lgtm
